### PR TITLE
Add vector labels

### PR DESCRIPTION
**Pull request by paritosh-08**
## TODO

- [x] Write a description
- [x] Add tests

## Description
This PR introduces new metric types:

- `CounterVector`
- `GuageVector`
- `HistogramVector`

The working of all of these is very similar. A vector metric of type `X` is nothing but a map of label and metric of type `X` registered under the same metric name. This allows us to add new labels for a given metric name during the runtime.

### How does this work?
We use the new `UncheckedDynamicSamplingGroup` for registering the vector metric and we modify the map (of label and metric) if a new label is reported during the runtime. See `createCounterVector` for example:
```hs
createCounterVector
  :: forall metrics name help labels.
      (KnownSymbol name, KnownSymbol help, ToLabels labels, Ord labels)
  => metrics name help 'CounterType labels -- ^ Metric class
  -> Store metrics -- ^ Metric store
  -> (Map labels Double -> [(labels, Double)]) -- ^ Hook for modifying the labels (eg. capping the number of labels exposed)
  -> IO (CounterVector labels)
createCounterVector metricConstructor store f = do
  counterVector <- CounterVector.new
  registerPermanently store $
    let group =
          UncheckedDynamicSamplingGroup ::> (metricConstructor, f)
     in registerUncheckedDynamicGroup
          group
          (CounterVector.read counterVector)
  pure counterVector
```

### "Capping" the number of labels
Notice the third argument in the function `createCounterVector`:
``` hs
f :: Map labels Double -> [(labels, Double)]
```
This hook is used while sampling to get the labels and their values. So, this can be used to "cap" the number of labels exposed while sampling. However, notice that we are storing all the metric labels and their values in memory, but we expose a limited set of labels based on the hook (`f`) during sampling.

**Note**: If you don't want to "cap" the lables, use [Map.toList](https://hackage.haskell.org/package/containers-0.4.0.0/docs/Data-Map.html#v:toList).

### CODE REVIEW

**Review by awjchen — COMMENTED** [review]
Nice work @paritosh-08! Thanks especially for the haddocks. This implementation of "metric vectors" will meet our needs.

Though, I've been thinking about it, and I think we could make the implementation simpler and more general if we implemented each metric vector type as a specialization of a generic vector type `newtype MetricVector label metric = C (IORef (Map label metric))` with these three operations:
1. `new :: IO (MetricVector label metric)`,
1. `getMetric :: MetricVector label metric -> label -> IO metric` (initializes and returns a new metric if not found), and
1. `getMetrics :: MetricVector label metric -> IO (Map label metric)`.

Then, `CounterVector` would be `newtype CounterVector label = C (MetricVector label Counter)`. To update the value of a specific `Counter` in such a `CounterVector`, one would first get the specific via `getCounter`, then use a function from the `Counter` interface to update it.

Basically, I'm suggesting that we make our metric vectors like more [`MetricVec` in the Go Prometheus client library](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#MetricVec).

Deduplicating the metric vector implementations and reducing the size of their APIs would allow us to really narrow the scope of our tests to just the underlying vector implementation. Right now we have to test each of the three types of metric vectors, plus their wrappers over the APIs of the individual metric type.

What do you think?

File: System/Metrics/Prometheus.hs
Comment (paritosh-08) [reply System/Metrics/Prometheus.hs]: Thanks Alex :raised_hands:

File: System/Metrics/Prometheus.hs
Comment (paritosh-08) [reply System/Metrics/Prometheus.hs]: Done, thanks Alex :raised_hands:

File: test/VectorMetrics.hs
Comment (paritosh-08) [reply test/VectorMetrics.hs:46]: Actually, I just followed what we were doing for other tests which involved sampling. However, using the `Sample` type is not super clean:
```hs
      -- ...
      CounterVector.add dynamicCounter (M.singleton "label_name_1" "label value 1") 10
      CounterVector.add dynamicCounter (M.singleton "label_name_2" "label value 1") 10

      prometheusSample1 <- sampleAll store

      let
        labelValueShouldBe :: Map.Map Labels Value -> Labels -> Double -> IO ()
        labelValueShouldBe prometheusSample labels value = do
          case Map.lookup labels prometheusSample of
            Nothing -> expectationFailure $ "label not found: " <> show labels
            Just (Counter value') -> value' `shouldBe` value
            Just (Gauge value') -> value' `shouldBe` value
            Just (Histogram _) -> expectationFailure $ "label is a histogram: " <> show labels
      case Map.lookup "my_dynamic_counter" prometheusSample1 of
        Nothing -> expectationFailure "my_dynamic_counter not found"
        Just (_, metricSample) -> do
          labelValueShouldBe metricSample (M.singleton "label_name_1" "label value 1") 10
          labelValueShouldBe metricSample (M.singleton "label_name_2" "label value 1") 10

      CounterVector.add dynamicCounter (M.singleton "label_name_2" "label value 1") 10
      CounterVector.add dynamicCounter (M.singleton "label_name_2" "label value 2") 10
      CounterVector.add dynamicCounter (M.singleton "label_name_3" "label value 1") 10

      prometheusSample2 <- sampleAll store

      case Map.lookup "my_dynamic_counter" prometheusSample2 of
        Nothing -> expectationFailure "my_dynamic_counter not found"
        Just (_, metricSample) -> do
          labelValueShouldBe metricSample (M.singleton "label_name_1" "label value 1") 10
          labelValueShouldBe metricSample (M.singleton "label_name_2" "label value 1") 20
          labelValueShouldBe metricSample (M.singleton "label_name_2" "label value 2") 10
          labelValueShouldBe metricSample (M.singleton "label_name_3" "label value 1") 10
      -- ...
```

I don't like the above approach; however, we can format the Prometheus exposition to make it more readable:
```hs
      -- ...
      GaugeVector.set dynamicGauge (M.singleton "label_name_1" "label value 1") 10
      GaugeVector.set dynamicGauge (M.singleton "label_name_2" "label value 1") 10

      prometheusSample1 <-
        BB.toLazyByteString . sampleToPrometheus <$> sampleAll store

      shouldBe
        prometheusSample1
          "# HELP my_dynamic_gauge Example gauge with dynamic labels\n\
          \# TYPE my_dynamic_gauge gauge\n\
          \my_dynamic_gauge{label_name_1=\"label value 1\"} 10.0\n\
          \my_dynamic_gauge{label_name_2=\"label value 1\"} 10.0"
      -- ...
```

**Review by awjchen — APPROVED** [review]
Looks good to me, thanks @paritosh-08!

We could further simplify the `CounterVector` by reducing its API to basically just a lookup function that returns a single `Counter` (initializing if necessary) and having the user operate on that single `Counter`. (And similarly for `GaugeVector` and `HistogramVector`.) But I suppose we could hold off on that change until we actually need to store references to individual `Counters` of a `CounterVector`.

So let's squash and merge this PR, once we make the below tweaks to the haddocks.

**Pull request merged**
